pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go

```diff
@@ -294,56 +299,55 @@ func getPodIPs(p *v1.Pod) []v1.PodIP {
 // See https://kubernetes.io/docs/concepts/services-networking/service/#services-without-selectors for more info.
 // For selector service, we query by the selector elsewhere, so this only handles the services that are NOT already found
 // by a selector.
+// For EndpointSlices that happen to point to the same IP as the pod, but are not directly bound to the pod (via TargetRef),
+// we ignore them here. These will produce a model.Workload directly from the EndpointSlice, but with limited information;
+// we do not implicitly merge a Pod with an EndpointSlice just based on IP.
 func (a *index) matchingServicesWithoutSelectors(
 	ctx krt.HandlerContext,
 	p *v1.Pod,
 	alreadyMatchingServices []model.ServiceInfo,
 	workloadServices krt.Collection[model.ServiceInfo],
 	endpointSlices krt.Collection[*discovery.EndpointSlice],
-	endpointSlicesAddressIndex krt.Index[string, *discovery.EndpointSlice],
+	endpointSlicesAddressIndex krt.Index[TargetRef, *discovery.EndpointSlice],
 ) []model.ServiceInfo {
-	k8sPodIPs := getPodIPs(p)
-	if len(k8sPodIPs) == 0 {
-		return nil
-	}
 	var res []model.ServiceInfo
 	// Build out our set of already-matched services to avoid double-selecting a service
 	seen := sets.NewWithLength[string](len(alreadyMatchingServices))
 	for _, s := range alreadyMatchingServices {
 		seen.Insert(s.Hostname)
 	}
-	for _, ip := range k8sPodIPs {
-		// For each IP, find any endpointSlices referencing it.
-		matchedSlices := krt.Fetch(ctx, endpointSlices, krt.FilterIndex(endpointSlicesAddressIndex, ip.IP))
-		for _, es := range matchedSlices {
-			if es.AddressType == discovery.AddressTypeFQDN {
-				// Currently we do not support FQDN.
-				continue
-			}
-			serviceName, f := es.Labels[discovery.LabelServiceName]
-			if !f {
-				// Not for a service; we don't care about it.
-				continue
-			}
-			hostname := string(kube.ServiceHostname(serviceName, es.Namespace, a.DomainSuffix))
-			if seen.Contains(hostname) {
-				// We already know about this service
-				continue
-			}
-			// This pod is included in the EndpointSlice. We need to fetch the Service object for it, by key.
-			serviceKey := es.Namespace + "/" + hostname
-			svcs := krt.Fetch(ctx, workloadServices, krt.FilterKey(serviceKey), krt.FilterGeneric(func(a any) bool {
-				// Only find Service, not Service Entry
-				return a.(model.ServiceInfo).Source == kind.Service
-			}))
-			if len(svcs) == 0 {
-				// no service found
-				continue
-			}
-			// There SHOULD only be one. This is only for `Service` which has unique hostnames.
-			svc := svcs[0]
-			res = append(res, svc)
+	tr := TargetRef{
```

Comment:
As discussed in the description, move from a lookup by IP to a lookup by targetRef